### ISSUE TITLE
Implement Population-Based Training for Hyperparameter Optimization

### ISSUE THREAD

**Issue by BY571** [TorchTrade/torchtrade#52]
## Problem

Single PPO agents with fixed hyperparameters **prematurely converge** to suboptimal strategies.

**Evidence:**
- All 10 reward function iterations converged to Action 6 (87% usage)
- Single entropy coefficient (0.01) insufficient for 19-action space
- No mechanism to escape local optima once value function converges

---

## Proposed Solution: Population-Based Training (PBT)

Train **multiple agents simultaneously** with different hyperparameters, then evolve the population by replacing poor performers with mutations of top performers.

### How PBT Works

```
Step 1: Initialize population (N=10 agents)
  Agent 1: lr=1e-5, entropy=0.01, ...
  Agent 2: lr=3e-5, entropy=0.02, ...
  Agent 3: lr=1e-4, entropy=0.05, ...
  ...

Step 2: Train all agents for T steps (e.g., 10k frames)

Step 3: Evaluate performance
  Rank agents by test Sharpe ratio

Step 4: Exploit + Explore
  - Bottom 20% copy weights from top 20%
  - Mutate hyperparameters (±20% perturbation)
  
Step 5: Repeat from Step 2
```

---

### Population Manager

```python
@dataclass
class AgentConfig:
    """Hyperparameters for a single agent."""
    learning_rate: float
    entropy_coef: float
    clip_epsilon: float
    gamma: float
    gae_lambda: float
    agent_id: int

class PopulationManager:
    """Manages population of agents with different hyperparameters."""
    
    def __init__(
        self,
        population_size: int = 10,
        exploit_top_k: int = 2,
        mutation_rate: float = 0.2,
    ):
        self.population_size = population_size
        self.exploit_top_k = exploit_top_k
        self.mutation_rate = mutation_rate
        
        # Initialize population with diverse hyperparameters
        self.agents = [
            self._create_agent(i) for i in range(population_size)
        ]
    
    def _create_agent(self, agent_id: int) -> AgentConfig:
        """Create agent with random hyperparameters."""
        return AgentConfig(
            learning_rate=np.random.uniform(1e-5, 1e-4),
            entropy_coef=np.random.uniform(0.01, 0.1),
            clip_epsilon=np.random.uniform(0.1, 0.3),
            gamma=np.random.uniform(0.9, 0.99),
            gae_lambda=np.random.uniform(0.9, 0.99),
            agent_id=agent_id,
        )
    
    def exploit_and_explore(
        self,
        agents: List[AgentConfig],
        performances: List[float],
    ) -> List[AgentConfig]:
        """Replace bottom performers with mutated top performers."""
        # Sort by performance
        sorted_idx = np.argsort(performances)[::-1]
        
        # Identify top and bottom performers
        top_k_idx = sorted_idx[:self.exploit_top_k]
        bottom_k_idx = sorted_idx[-self.exploit_top_k:]
        
        # Replace bottom with mutated top
        for bottom_idx in bottom_k_idx:
            # Copy random top performer
            top_idx = np.random.choice(top_k_idx)
            agents[bottom_idx] = self._mutate(agents[top_idx], bottom_idx)
        
        return agents
    
    def _mutate(self, agent: AgentConfig, new_id: int) -> AgentConfig:
        """Mutate hyperparameters."""
        mutated = AgentConfig(
            learning_rate=agent.learning_rate * np.random.uniform(0.8, 1.2),
            entropy_coef=agent.entropy_coef * np.random.uniform(0.8, 1.2),
            clip_epsilon=agent.clip_epsilon * np.random.uniform(0.8, 1.2),
            gamma=np.clip(agent.gamma * np.random.uniform(0.99, 1.01), 0.9, 0.99),
            gae_lambda=np.clip(agent.gae_lambda * np.random.uniform(0.99, 1.01), 0.9, 0.99),
            agent_id=new_id,
        )
        return mutated
```

### Training Loop

```python
def train_population(
    train_env,
    eval_env,
    population_manager: PopulationManager,
    total_generations: int = 10,
    frames_per_generation: int = 10_000,
):
    """Train population with PBT."""
    
    for generation in range(total_generations):
        print(f"\n=== Generation {generation} ===")
        
        # Train all agents for frames_per_generation
        performances = []
        for agent_config in population_manager.agents:
            # Create agent with these hyperparameters
            actor, critic = create_ppo_models(agent_config)
            
            # Train for frames_per_generation
            collector = create_collector(train_env, actor, agent_config)
            train_agent(collector, actor, critic, agent_config, frames_per_generation)
            
            # Evaluate on test set
            test_sharpe = evaluate(eval_env, actor)
            performances.append(test_sharpe)
            
            print(f"Agent {agent_config.agent_id}: Sharpe={test_sharpe:.2f}, lr={agent_config.learning_rate:.2e}")
        
        # Exploit + Explore
        population_manager.agents = population_manager.exploit_and_explore(
            population_manager.agents,
            performances
        )
        
        # Log population statistics
        log_population_stats(population_manager.agents, performances, generation)
```

---

### Success Metrics

| Metric | Single Agent | PBT Population |
|--------|-------------|----------------|
| Best agent Sharpe | -5.77 | > 0.0 |
| Action entropy | 0.48 | > 1.0 (for best agent) |
| Convergence time | N/A | Faster (parallel exploration) |
| Hyperparameter sensitivity | High | Low (auto-tuned) |

### Advantages

1. **Automatic hyperparameter tuning**: No manual grid search
2. **Escape local optima**: Poor performers reset to good configurations
3. **Diverse exploration**: Multiple agents try different strategies simultaneously
4. **Adaptive**: Hyperparameters evolve as training progresses
5. **Parallel efficiency**: All agents train in parallel

---

## Hyperparameter Search Space

```python
SEARCH_SPACE = {
    "learning_rate": [1e-5, 1e-4],          # Log-uniform
    "entropy_coef": [0.01, 0.1],            # Log-uniform
    "clip_epsilon": [0.1, 0.3],             # Uniform
    "gamma": [0.90, 0.99],                  # Uniform
    "gae_lambda": [0.90, 0.99],             # Uniform
    "mini_batch_size": [256, 512, 1024],    # Discrete
    "ppo_epochs": [2, 3, 5],                # Discrete
}
```

**Critical hypothesis**: Agents with **higher entropy_coef** (0.05-0.1) will outperform baseline (0.01) by discovering more strategies.

---

### Week 1: Core PBT Infrastructure
- [ ] Implement `PopulationManager` class
- [ ] Implement `exploit_and_explore` logic
- [ ] Add parallel training support (multiprocessing)
- [ ] Add population logging and visualization

### Week 2: Integration with PPO
- [ ] Modify training script to support multiple agents
- [ ] Add checkpoint saving/loading per agent
- [ ] Implement evaluation protocol (test set Sharpe ratio)
- [ ] Add mutation strategies (resample vs perturb)

### Week 3: Experiments
- [ ] Baseline: Single agent (entropy=0.01)
- [ ] PBT: 10 agents, 10 generations, 10k frames/generation
- [ ] Analyze best agent hyperparameters
- [ ] Compare action distributions across population
- [ ] Plot hyperparameter evolution over generations

---

## Configuration Example

```yaml
# config.yaml
pbt:
  enabled: true
  population_size: 10
  generations: 10
  frames_per_generation: 10000
  
  exploit:
    top_k: 2        # Copy top 2 performers
    bottom_k: 2     # Replace bottom 2
  
  explore:
    mutation_rate: 0.2    # ±20% perturbation
    mutation_type: "perturb"  # or "resample"
  
  search_space:
    learning_rate: [1e-5, 1e-4]
    entropy_coef: [0.01, 0.1]
    clip_epsilon: [0.1, 0.3]
    gamma: [0.9, 0.99]
```

---

## Monitoring & Visualization

Track these metrics per generation:

```python
metrics = {
    # Performance
    "pbt/best_sharpe": max(performances),
    "pbt/mean_sharpe": np.mean(performances),
    "pbt/worst_sharpe": min(performances),
    
    # Hyperparameters
    "pbt/mean_entropy_coef": np.mean([a.entropy_coef for a in agents]),
    "pbt/std_entropy_coef": np.std([a.entropy_coef for a in agents]),
    "pbt/mean_lr": np.mean([a.learning_rate for a in agents]),
    
    # Diversity
    "pbt/action_entropy_best_agent": compute_entropy(best_agent),
    "pbt/unique_strategies": count_unique_strategies(agents),
}
```

**Visualizations:**
- Hyperparameter evolution over generations (scatter plots)
- Performance distribution per generation (box plots)
- Best agent action distribution vs baseline

---

## Computational Requirements

**Single agent baseline:**
```
100k frames × 1 agent = 100k total frames
Training time: ~2 hours
```

**PBT:**
```
10k frames × 10 agents × 10 generations = 1M total frames
BUT: Parallelizable across agents
Wall-clock time: ~2-3 hours (with 10 GPUs)
                ~20 hours (with 1 GPU, sequential)
```

**Recommendation**: Use GPU cluster or cloud instances for parallel training.

---

## References

**Population-Based Training:**
- Jaderberg et al. (2017) - "Population Based Training of Neural Networks"
- Parker-Holder et al. (2020) - "Effective Diversity in Population Based Reinforcement Learning"

**Applications:**
- OpenAI Five (2018) - Used PBT for Dota 2
- AlphaStar (2019) - PBT for StarCraft II
- FTW agent (2019) - PBT for Capture the Flag

---

### Variant A: Tournament Selection
Instead of top-k, use tournament selection:
```python
def tournament_selection(agents, performances, k=3):
    # Randomly select k agents
    # Best one survives, others replaced
```

### Variant B: Gradual Mutation
Start with large mutations, decay over time:
```python
mutation_rate = 0.5 * (0.5 ** (generation / 5))  # 0.5 → 0.1 over 10 gens
```

### Variant C: PBT + Curriculum
Combine with curriculum learning (#49):
```python
# All agents follow same curriculum
# But explore different hyperparameters within each phase
```

---

## Related Issues

- #48 - Curiosity (can be added to search space)
- #49 - Curriculum (can combine with PBT)
- #50 - Entropy scheduling (PBT auto-tunes this)
- #51 - Hierarchical policy (can be evolved via PBT)
